ee/server/service/software_installers.go

```diff
@@ -115,6 +116,257 @@ func preProcessUninstallScript(payload *fleet.UploadSoftwareInstallerPayload) {
+	if err := svc.authz.Authorize(ctx, &fleet.SoftwareInstaller{TeamID: payload.TeamID}, fleet.ActionWrite); err != nil {
+		return nil, err
+	}
+
+	vc, ok := viewer.FromContext(ctx)
+	if !ok {
+		return nil, fleet.ErrNoContext
+	}
+	payload.UserID = vc.UserID()
+
+	if payload.TeamID == nil {
+		return nil, &fleet.BadRequestError{Message: "team_id is required; enter 0 for no team"}
+	}
+
+	var teamName *string
+	if *payload.TeamID != 0 {
+		t, err := svc.ds.Team(ctx, *payload.TeamID)
+		if err != nil {
+			return nil, err
+		}
+		teamName = &t.Name
+	}
+
+	// get software by ID, fail if it does not exist or does not have an installer
+	software, err := svc.ds.SoftwareTitleByID(ctx, titleID, payload.TeamID, fleet.TeamFilter{
+		User:            vc.User,
+		IncludeObserver: true,
+	})
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "getting software title by id")
+	}
+
+	// TODO when we start supporting multiple installers per title X team, need to rework how we determine installer to edit
+	if software.SoftwareInstallersCount != 1 {
+		return nil, &fleet.BadRequestError{
+			Message: "There are no software installers defined yet for this title and team. Please add an installer instead of attempting to edit.",
+		}
+	}
+
+	installer, err := svc.ds.GetSoftwareInstallerMetadataByTeamAndTitleID(ctx, payload.TeamID, titleID, true)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "getting existing installer")
+	}
+
+	if payload.SelfService == nil && payload.InstallerFile == nil && payload.PreInstallQuery == nil &&
+		payload.InstallScript == nil && payload.PostInstallScript == nil && payload.UninstallScript == nil {
+		return installer, nil // no payload, noop
+	}
+
+	payload.InstallerID = installer.InstallerID
+	var dirty []string
+
+	if payload.SelfService != nil && *payload.SelfService != installer.SelfService {
+		dirty = append(dirty, "SelfService")
+	}
+
+	activity := fleet.ActivityTypeEditedSoftware{
+		SoftwareTitle: installer.SoftwareTitle,
+		TeamName:      teamName,
+		TeamID:        payload.TeamID,
+		SelfService:   installer.SelfService,
+	}
+
+	var payloadForNewInstallerFile *fleet.UploadSoftwareInstallerPayload
+	if payload.InstallerFile != nil || payload.PreInstallQuery != nil ||
+		payload.InstallScript != nil || payload.PostInstallScript != nil || payload.UninstallScript != nil {
+
+		// pull existing installer to calculate metadata for defaulting/validation
+
+		// check if the installer exists in the store
+		exists, err := svc.softwareInstallStore.Exists(ctx, installer.StorageID)
+		if err != nil {
+			return nil, ctxerr.Wrap(ctx, err, "checking if installer exists")
+		}
+		if !exists {
+			return nil, ctxerr.Wrap(ctx, notFoundError{}, "does not exist in software installer store")
+		}
+
+		// get the installer from the store
+		existingInstallerFile, _, err := svc.softwareInstallStore.Get(ctx, installer.StorageID)
```

Comment:
seems exessive to pull the installer just to extract metadata.  guessing we're not persisting enough data in `software_installers`?

Reply:
We're close, but are missing the following:

1. Package IDs are missing from the `SoftwareInstaller` struct, though they're present in the DB. I'll see if I can add those into the struct, and populate them from the DB.
2. Magic-derived estension on installer. This is distinct from the extension of the installer filename, and we don't have a column for this. [See this comment for background.](https://github.com/fleetdm/fleet/pull/21892#issuecomment-2337855820).

Right now we already have one item this release (uninstall script generation) that pulls every installer to populate something in the database, and does so asynchronously, so I'm a little reluctant to throw another in as a separate job + DB migration. Maybe we can add extension as its own column, to be built out as part of that migration script though? If that happens I can pull everything I need from the DB once the process completes, and can fall back to filename based while the process is running.

Reply:
Okay, I'll get extension added to the DB as a PR on https://github.com/fleetdm/fleet/pull/21892.

Reply:
Got this cleaned up; with the latest DB changes (which will all be done as part of the uninstall PR) we can just look at the DB for what we need from the old installer.